src/Control/Carrier/Diagnostics.hs

```diff
@@ -75,6 +82,29 @@ renderFailureBundle FailureBundle {..} =
 renderWarnings :: [SomeDiagnostic] -> Doc AnsiStyle
 renderWarnings = align . vsep . map renderSomeDiagnostic
 
+logResultWarnings :: Has Logger sig m => ResultBundle a -> m a
+logResultWarnings ResultBundle {..} = logWarn (renderWarnings resultWarnings) $> resultValue
+
+logErrorBundle :: Has Logger sig m => FailureBundle -> m ()
+logErrorBundle = logError . renderFailureBundle
+
+-- | Run a Diagnostic effect into a logger, using the default error/warning renderers.
+logDiagnostic :: Has Logger sig m => DiagnosticsC m a -> m (Maybe a)
+logDiagnostic diag = do
+  result <- runDiagnostics diag
+  case result of
+    Left failure -> logErrorBundle failure >> pure Nothing
+    Right success -> Just <$> logResultWarnings success
+
+-- | Run a void Diagnostic effect into a logger, using the default error/warning renderers.
+-- | Exits with non-zero if the result is a failure.
+-- | Useful for setting up diagnostics from CLI entry points.
+logDiagnostic_ :: (Has (Lift IO) sig m, Has Logger sig m) => DiagnosticsC m () -> m ()
```

Comment:
I like this. I wonder if there's a name we can use to better signal "this will immediately terminate the program when it encounters a failure"

Reply:
I had thought about `logDiagnostic!` as an alternative

Reply:
For now, `Diag.logWithExit_` will do.